cs3api4lab/cs3_file_api.py

```diff
@@ -0,0 +1,237 @@
+        except requests.exceptions.RequestException as e:
+            self.log.error('msg="Exception when downloading file from Reva" reason="%s"' % e)
+            yield IOError(e)
+
+        time_end = time.time()
+        data = file_get.content
+
+        if file_get.status_code != http.HTTPStatus.OK:
+            self.log.error('msg="Error downloading file from Reva" code="%d" reason="%s"' % (file_get.status_code, file_get.reason))
+            yield IOError(file_get.reason)
+        else:
+            self.log.info('msg="File open for read" filepath="%s" elapsedTimems="%.1f"' % (filepath, (time_end - time_start) * 1000))
+            for i in range(0, len(data), self.chunksize):
+                yield data[i:i + self.chunksize]
+
+    def write_file(self, endpoint, filepath, userid, content, noversion=0):
+        """
+        Write a file using the given userid as access token. The entire content is written
+        and any pre-existing file is deleted (or moved to the previous version if supported).
+        The noversion flag is currently not supported.
+        """
+
+        #
+        # Prepare endpoint
+        #
+        time_start = time.time()
+
+        reference = self._cs3_reference(endpoint, filepath)
+
+        req = cs3sp.InitiateFileUploadRequest(ref=reference)
+        initfileuploadres = self.cs3_stub.InitiateFileUpload(request=req, metadata=[('x-access-token', self._authenticate(userid))])
+
+        if initfileuploadres.status.code != cs3code.CODE_OK:
+            self.log.debug('msg="Failed to initiateFileUpload on write" filepath="%s" reason="%s"' % \
+                                  (filepath, initfileuploadres.status.message))
+            raise IOError(initfileuploadres.status.message)
+
+        self.log.debug('msg="writefile: InitiateFileUploadRes returned" endpoint="%s"' % initfileuploadres.upload_endpoint)
+
+        #
+        # Upload
+        #
+        try:
+            # TODO: Use tus client instead of PUT
+            headers = {
+                'Tus-Resumable': '1.0.0',
+                'x-access-token': self._authenticate(userid)
+            }
+            putres = requests.put(url=initfileuploadres.upload_endpoint, data=content, headers=headers)
+
+        except requests.exceptions.RequestException as e:
+            self.log.error('msg="Exception when uploading file to Reva" reason="%s"' % e)
+            raise IOError(e)
+
+        time_end = time.time()
+
+        if putres.status_code != http.HTTPStatus.OK:
+            self.log.error('msg="Error uploading file to Reva" code="%d" reason="%s"' % (putres.status_code, putres.reason))
+            raise IOError(putres.reason)
+
+        self.log.info('msg="File open for write" filepath="%s" elapsedTimems="%.1f"' % (filepath, (time_end - time_start) * 1000))
+
+    def remove_file(self, endpoint, filepath, userid, force=0):
```

Comment:
We also remove folders. So maybe a more generic name would be better.
Why do we need the force? Does REVA prevent deletion of non empty folders or smth like that?

Reply:
A generic name is good. "Force" attribute is using in WOPI at xrootiface. This attribute is for compatibility, we can remove because Reva does not support it.